File: 프로그래머스/3/42898. 등굣길/등굣길.js
function solution(m, n, puddles) {
    const dp = new Array(m+1).fill(0).map(() => new Array(n+1).fill(0));
    puddles.forEach(([x, y]) => dp[x][y] = -1);
    
    dp[1][1] = 1;
    
    for (let i = 1; i <= m; i++) {
        for (let j = 1; j <= n; j++) {
            if (dp[i][j] === -1) continue;
            dp[i][j] += Math.max(dp[i-1][j], 0) + Math.max(dp[i][j-1], 0);
            dp[i][j] %= 1000000007;
        }
    }
    
    return Math.max(dp[m][n], 0);
}

File: 프로그래머스/3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    dp = [[0] * (n+1) for _ in range(m+1)]
    
    for x, y in puddles:
        dp[x][y] = -1
    
    dp[1][1] = 1
    
    for i in range(1, m+1):
        for j in range(1, n+1):
            if dp[i][j] == -1:
                continue
            dp[i][j] += max(dp[i-1][j], 0) + max(dp[i][j-1], 0)
            dp[i][j] %= 1000000007
    
    return dp[m][n]
